pkg/util/evicting_queue.go

```diff
@@ -0,0 +1,69 @@
+package util
+
+import (
+	"errors"
+	"sync"
+)
+
+type EvictingQueue struct {
+	sync.RWMutex
+
+	capacity int
+	entries  []interface{}
+	onEvict  func()
+}
+
+func NewEvictingQueue(capacity int, onEvict func()) (*EvictingQueue, error) {
+	if capacity <= 0 {
+		// a queue of 0 (or smaller) capacity is invalid
+		return nil, errors.New("queue cannot have a zero or negative capacity")
+	}
+
+	return &EvictingQueue{
+		capacity: capacity,
+		onEvict:  onEvict,
+		entries:  make([]interface{}, 0, capacity),
+	}, nil
+}
+
+func (q *EvictingQueue) Append(entry interface{}) {
+	q.Lock()
+	defer q.Unlock()
+
+	if len(q.entries) >= q.capacity {
+		q.evictOldest()
```

Comment:
nit: What you're doing here looks good, but we could be a bit safer in the case the entries somehow become significantly larger than the capacity.
```suggestion
	q.onEvict()
	start := (len(q.entries)-q.Capacity()) + 1
	q.entries = append(q.entries[:0], q.entries[start:]...)
```